frappe/model/db_query.py

```diff
@@ -808,7 +807,10 @@ def prepare_filter_condition(self, f):
 			df = meta.get("fields", {"fieldname": f.fieldname})
 			df = df[0] if df else None
 
-			if df and df.fieldtype in ("Check", "Float", "Int", "Currency", "Percent"):
+			if df and (
+				df.fieldtype in ("Check", "Float", "Int", "Currency", "Percent")
+				or getattr(df, "not_nullable", False)
+			):
```

Comment:
Implement this for `in` and `not in` operator too and for NestedSet operators too.

Reply:
Also check if `can_be_null` is again being set by some other heuristics down below (e.g. `not set`)